`write_ndjson` produces incorrect data that does not match dataframe

### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import json

import polars as pl
import polars.selectors as pls


def transform(df: pl.DataFrame):
    aggregated = (
        df.with_columns(feature="@" + pl.col("feature"))
        .pivot(index=["identifier", "day"], on="feature", values="value")
        .lazy()
        .sort(by=["identifier", "day"])
        .fill_null(0)
        .group_by("identifier")
        .agg(pls.all())
        .collect()
    )
    return aggregated.with_columns(
        **{
            name: pl.when(pl.col(name).list.sum() == 0)
            .then(None)
            .otherwise(pl.col(name))
            for name in pls.expand_selector(aggregated, pls.starts_with("@"))
        }
    )


def check_for_error(df: pl.DataFrame):
    polar_lines = [json.loads(l) for l in df.write_ndjson().strip().split("\n")]
    pandas_lines = [
        json.loads(l)
        for l in df.to_pandas()
        .to_json(orient="records", lines=True)
        .strip()
        .split("\n")
    ]

    polars_map = {line["identifier"]: line for line in polar_lines}
    pandas_map = {line["identifier"]: line for line in pandas_lines}

    for key in pandas_map:
        a_row = pandas_map[key]
        b_row = polars_map[key]
        for k in a_row:
            if k.startswith("@") and a_row[k] != b_row[k]:
                print("ND JSON inconsistent between polars and pandas")
                print(f"  Identifier: {key}")
                print(f"  Column: {k}")
                print(f"  Polars: {b_row[k]}")
                print(f"  Pandas: {a_row[k]}")
                print()


raw_data = {
    "identifier": ["0", "0", "0", "0", "0", "1", "1", "1", "1"],
    "day": [0, 1, 2, 3, 4, 1, 2, 3, 4],
    "feature": ["c", "c", "a", "c", "a", "b", "b", "b", "b"],
    "value": [74, None, None, 98, None, None, 60, None, None],
}

print("Polars Version: ", pl.__version__)
df = pl.DataFrame(raw_data)
aggregated = transform(df)
check_for_error(transform(df))

# Polars Version:  1.8.2
# ND JSON inconsistent between polars and pandas
#   Identifier: 0
#   Column: @c
#   Polars: [0, 0, 0, 0, 74]
#   Pandas: [74, 0, 0, 98, 0]
```


### Log output

```shell
keys/aggregates are not partitionable: running default HASH AGGREGATION
keys/aggregates are not partitionable: running default HASH AGGREGATION
```


### Issue description

In very specific situations `write_ndjson` produces records that are simply incorrect and do not reflect the data one would see by just printing the data frame. The example provided checks for this error by comparing records produced directly from `write_ndjson` and those produced by going from `polars` to `pandas` and then to ndjson. The way in which `polars` encodes the errant row is not fully consistent between runs. The error does seem to have something to do with:

```python
    return aggregated.with_columns(
        **{
            name: pl.when(pl.col(name).list.sum() == 0)
            .then(None)
            .otherwise(pl.col(name))
            for name in pls.expand_selector(aggregated, pls.starts_with("@"))
        }
    )
```

And specifically the statement `.then(None)`. In my testing if that is changed to any other value I can't recreate the error.

### Expected behavior

The output of polars `write_ndjson` should be consistent with the data presented when printing the data frame and should also be consistent with the same data produced via the `polars -> pandas -> ndjson` path.

### Installed versions

<details>

```
--------Version info---------
Polars:              1.8.2
Index type:          UInt32
Platform:            Linux-5.15.0-122-generic-x86_64-with-glibc2.31
Python:              3.12.2 (main, Apr  1 2024, 14:08:15) [GCC 9.4.0]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          <not installed>
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               <not installed>
gevent               <not installed>
great_tables         <not installed>
matplotlib           <not installed>
nest_asyncio         <not installed>
numpy                2.1.1
openpyxl             <not installed>
pandas               2.2.3
pyarrow              17.0.0
pydantic             <not installed>
pyiceberg            <not installed>
sqlalchemy           <not installed>
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

I think this is still a problem - I just ran into it again.

Simpler repro.

```python
import io
import polars as pl

df = pl.DataFrame({"x": [1, 2], "y": [None, 3]})

output_file = io.BytesIO()

query = (
    df.group_by("x", maintain_order=True)
      .all()
      .select(
          pl.when(pl.col("y").list.sum() > 0).then("y")
      )
)

eager = query.write_ndjson().encode()

query.lazy().sink_ndjson(output_file)
lazy = output_file.getvalue()

assert eager == lazy
# AssertionError: 
```
```python
b'{"y":null}\n{"y":[null]}\n'  # eager
b'{"y":null}\n{"y":[3]}\n'     # lazy
```

--- a/crates/polars-json/src/json/write/serialize.rs
+++ b/crates/polars-json/src/json/write/serialize.rs
@@ -240,8 +240,15 @@ fn list_serializer<'a, O: Offset>(
     let end = offsets.last().unwrap().to_usize();
     let mut serializer = new_serializer(array.values().as_ref(), start, end - start);
 
+    let mut prev_offset = start;
     let f = move |offset: Option<&[O]>, buf: &mut Vec<u8>| {
         if let Some(offset) = offset {
+            if offset[0].to_usize() > prev_offset {
+                for _ in 0..(offset[0].to_usize() - prev_offset) {
+                    serializer.next().unwrap();
+                }
+            }
+
             let length = (offset[1] - offset[0]).to_usize();
             buf.push(b'[');
             let mut is_first_row = true;
@@ -253,6 +260,7 @@ fn list_serializer<'a, O: Offset>(
                 buf.extend(serializer.next().unwrap());
             }
             buf.push(b']');
+            prev_offset = offset[1].to_usize();
         } else {
             buf.extend(b"null");
         }

--- a/py-polars/tests/unit/io/test_json.py
+++ b/py-polars/tests/unit/io/test_json.py
@@ -563,3 +563,22 @@ def test_read_json_utf_8_sig_encoding() -> None:
     result = pl.read_json(json.dumps(data).encode("utf-8-sig"))
     expected = pl.DataFrame(data)
     assert_frame_equal(result, expected)
+
+
+def test_write_masked_out_list_22202() -> None:
+    df = pl.DataFrame({"x": [1, 2], "y": [None, 3]})
+
+    output_file = io.BytesIO()
+
+    query = (
+        df.group_by("x", maintain_order=True)
+        .all()
+        .select(pl.when(pl.col("y").list.sum() > 0).then("y"))
+    )
+
+    eager = query.write_ndjson().encode()
+
+    query.lazy().sink_ndjson(output_file)
+    lazy = output_file.getvalue()
+
+    assert eager == lazy

